### PR TITLE
MPP plugin: add expectedBy deps to "api" configuration

### PR DESCRIPTION
**Pull request by bingranl**
Currently, KotlinPlatformJvmPlugin and KotlinPlatformJsPlugin are
adding expectedBy deps to "compile" configuration which is deprecated
by Gradle. This PR fixes that by replacing "compile" with "api" which
is what we are doing in KotlinPlatformAndroidPlugin.

This PR also makes integration tests running with warning-mode=fail by
default and fixes most of the integration tests. For the remaining tests
to be fixed, we make them run with warning-mode=summary and will fix
them incrementally in following PRs.

### CODE REVIEW

File: libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BuildCacheRelocationIT.kt [hunk 1]
```diff
@@ -36,7 +36,6 @@ class BuildCacheRelocationIT : BaseGradleIT() {
     override fun defaultBuildOptions(): BuildOptions =
         super.defaultBuildOptions().copy(
             withBuildCache = true,
-            androidGradlePluginVersion = AGPVersion.v3_6_0,
```
Comment (bingranl) [inline libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BuildCacheRelocationIT.kt:39]: Currently we rely on` BuildOptions.androidGradlePluginVersion` to determine whether it is an android test and what AGP version it is using.

Set Agp version only for those android tests to allow other non-android test run with warning-mode=fail if possible.

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinMultiplatformPlugin.kt [hunk 2]
```diff
@@ -239,10 +239,6 @@ open class KotlinPlatformAndroidPlugin : KotlinPlatformImplementationPluginBase(
         super.apply(project)
     }
 
-    override fun configurationsForCommonModuleDependency(project: Project): List<Configuration> =
-        (project.configurations.findByName("api"))?.let(::listOf)
-            ?: super.configurationsForCommonModuleDependency(project) // older Android plugins don't have api/implementation configs
```
Comment (bingranl) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinMultiplatformPlugin.kt:244]: The oldest AGP version used in integration test is 3.1.0(released in March 2018) , so I assume it is the oldest one KGP support, which should have api/implementation configs.

File: libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BaseGradleIT.kt [hunk 3]
```diff
@@ -153,6 +153,20 @@ abstract class BaseGradleIT {
             return wrapper
         }
 
+        fun mightUpdateSettingsScript(wrapperVersion: String, settingsScript: File) {
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BaseGradleIT.kt:156]: Rename to "maybeUpdateSettingsScript"?

File: libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BaseGradleIT.kt [hunk 4]
```diff
@@ -153,6 +153,20 @@ abstract class BaseGradleIT {
...
+            // enableFeaturePreview("GRADLE_METADATA") is no longer needed when building with Gradle 5.4 or above
+            if (GradleVersion.version(wrapperVersion) > GradleVersion.version("5.3")) {
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BaseGradleIT.kt:158]: Change to "version>= 5.4".

File: libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinMultiplatformPlugin.kt [hunk 5]
```diff
@@ -56,7 +56,7 @@ open class KotlinPlatformImplementationPluginBase(platformName: String) : Kotlin
     private val commonProjects = arrayListOf<Project>()
 
     protected open fun configurationsForCommonModuleDependency(project: Project): List<Configuration> =
```
Comment (gavra0) [inline libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinMultiplatformPlugin.kt:58]: Make this one private.

File: libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BuildCacheRelocationIT.kt (hunk 1 above)
Comment (gavra0) [reply libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/BuildCacheRelocationIT.kt:39]: Do we need this change, as it seems that all tests in this class are running with `--warning-mode=fail`?

**Review by gavra0 — APPROVED** [review]

**Comment by h0tk3y** [comment]
@bingranl The `KotlinPlatformJvmPlugin` and `KotlinPlatformJsPlugin` are deprecated Kotlin 1.2.x plugins that we are going to drop. They are replaced by the Kotlin Multiplatform plugin. Do you have a use case for the deprecated plugins that doesn't allow you to migrate?

I'll review the other changes in the PR soon, too.

**Comment by bingranl** [comment]
@h0tk3y Good to know that those plugins are going to be deprecated. But we still have some integration tests using those plugins which prevents us running with` --warning-mode=fail` if we don't replacing "compile" with "api"

**Comment by bingranl** [comment]
Friendly ping, :) @h0tk3y Does it look good to you?

**Review by h0tk3y — APPROVED** [review]

**Comment by h0tk3y** [comment]
The changes look good to me. Let me run the tests on our CI.

**Comment by h0tk3y** [comment]
Thanks for the PR!

**Comment by bingranl** [comment]
@h0tk3y Thanks for reviewing and merging it!

**Pull request merged**